camel/benchmarks/gaia.py

```diff
@@ -0,0 +1,246 @@
+            )
+        if isinstance(level, str) and level == "all":
+            levels = [1, 2, 3]
+        elif isinstance(level, int):
+            levels = [level]
+        else:
+            levels = level
+        if any(_level not in [1, 2, 3] for _level in levels):
+            raise ValueError(
+                f"Invalid value for `level`: {level},"
+                " expected 1, 2, 3 or 'all'."
+            )
+
+        logger.info(f"Running benchmark on {on} set at levels {levels}.")
+        datas = [data for data in self._data[on] if data["Level"] in levels]
+        if randomize:
+            random.shuffle(datas)
+        if subset:
+            datas = datas[:subset]
+        logger.info(f"Number of tasks: {len(datas)}")
+        self._results = []
+        agent = self._inject(agent)
+        with open(self.save_to, "w") as f:
+            for task in tqdm(datas, desc="Running"):
+                if task["file_name"] != "":
+                    logger.info("Skipping task with file_name.")
+                    continue
+                msg = BaseMessage.make_user_message(
+                    role_name="User",
+                    content=task["Question"],
+                )
+                final_answer = task["Final answer"]
+                try:
+                    result = agent.step(msg)
+                    model_answer = self.get_final_answer(
+                        result.msgs[0].content
+                    )
+                    tool_calls = result.info.get("tool_calls", [])
+                    score = self.question_scorer(model_answer, final_answer)
+                    self._results.append(
+                        {
+                            "task_id": task["task_id"],
+                            "model_answer": model_answer,
+                            "ground_truth": final_answer,
+                            "tool_calls": [
+                                tool.model_dump() for tool in tool_calls
+                            ],
+                            "error": None,
+                            "score": int(score),
+                        }
+                    )
+                except Exception as e:
+                    logger.warning(
+                        f"Error in processing task: {task['task_id']}"
+                    )
+                    self._results.append(
+                        {
+                            "task_id": task["task_id"],
+                            "model_answer": "ERROR",
+                            "ground_truth": final_answer,
+                            "tool_calls": [],
+                            "error": str(e),
+                            "score": 0,
+                        }
+                    )
+                agent.reset()
+
+                self._results[-1]["history"] = self._current_history
+                self._current_history = []
+                f.write(json.dumps(self._results[-1], indent=2) + "\n")
+                f.flush()
```

Comment:
how about using temporary files for atomic writes, which could be more safe, unexpected behaviour during directly writing could break the file